### ISSUE TITLE
Personal access tokens

### ISSUE THREAD

**Issue by ssbb** [evonove/django-oauth-toolkit#257]
I tried to implement something like GitHub personal access tokens with scopes support. It's no problem to do it with custom model and bearer tokens, but when I want to limit scopes too, I think that is a good idea to manage this logic with existing OAuth library.

So I see two ways to do this:

1. Allow null values on `AccessToken` model for foreign key to `Application.
2. Make all models swappable and not only `Application`.

I am ready to prepare PR, but I would like to know which method is preferable. 

Thanks :)

**Comment by synasius** [comment]
Hi @ssbb,
first of all sorry for the delay and thanks for your time.

TLDR; Personal tokens are nice but this feature can be implemented with a few custom views without changing the underlying models.

The idea is nice but I think is outside the scope of DOT. However, I think that a Personal token manger can be built upon DOT without changing the underlying model for the following reasons:

> Make all models swappable and not only Application.

Swappable models are a pain. Maintaining migrations is hard, people always complain for corner cases, and are also hard to tests. I think that introducing the swappable Application model was an error, a big one, and I don't want to go down that road again.

> Allow null values on AccessToken model for foreign key to `Application.

Is this really necessary? This change should then be handled and tested in the code wherever application is used. I think that personal tokens should be bound to a default application, and that can be done in the view that creates the personal access token. Scopes should also be managed there: you can show a form to the user with the available scopes. Or am I missing something?
What do you think?

**Comment by ssbb** [comment]
Hi @synasius,
thanks for detailed response.

Yes, I have already implemented it through an internal "default" application.

Swappable models are really painful for maintainers, but easy to use for end users. I think it is more useful than OneToOne relations like we do it with auth.User before. On the other hand, it makes reusable applications far more complicated.

**Comment by pySilver** [comment]
@ssbb so you tied your "default" application to some kind of superadmin? Actually GitHub itself doing something like that – since all client_id issued for personal user are the same. I just hate the fact that Application needs a User, so one should have superadmin added just for this matter

**Comment by burnash** [comment]
@ssbb do you have any plans to share an example of the implementation? E.g. in a blog post or a gist. I think personal tokens is a good idea and the community might benefit from knowledge how to do it right.

**Comment by ssbb** [comment]
@burnash it's too easy even for blog post :)

You just creating internal User and Application. Then on "get token" you can create new tokens and attach them to this application.

**Comment by burnash** [comment]
@ssbb thanks for sharing. We used an internal user and application for [creating new tokens](https://github.com/NeuroVault/NeuroVault/blob/fd673f2f2129f6a3ccff0fc8aca73b6d9d75c345/neurovault/apps/users/views.py#L103-L118) in our app and so far it works just fine.
